test/gtest/ucp/test_ucp_sockaddr.cc

```diff
@@ -1117,62 +1132,72 @@ class test_ucp_sockaddr_protocols : public test_ucp_sockaddr {
         EXPECT_EQ(0,    recv_info.sender_tag);
 
         if (err_handling) {
-            if (recv_stop != NULL) {
-                recv_stop(arg);
+            if (recv_stop) {
+                disconnect(*this, receiver());
             }
 
-            /* If send request is NULL, it means that send operation
-             * completed immediately */
-            if (sreq != NULL) {
-                /* TODO: remove memory deregistration, when UCP
-                 * requests tracking is added */
-                ucp_request_t *req = static_cast<ucp_request_t*>
-                                     (sreq) - 1;
-                EXPECT_EQ(sender().ucph(),
-                          req->send.ep->worker->context);
-                ucp_request_memory_dereg(sender().ucph(),
-                                         req->send.datatype,
-                                         &req->send.state.dt, req);
-            }
+            sreq_mem_dereg(sreq);
 
-            if (recv_stop != NULL) {
+            if (recv_stop) {
                 sender().disconnect_nb(0, 0, UCP_EP_CLOSE_MODE_FORCE);
             } else {
-                send_stop(arg);
+                disconnect(*this, sender());
                 receiver().disconnect_nb(0, 0, UCP_EP_CLOSE_MODE_FORCE);
             }
         }
 
-        return ucp_tag_msg_recv_nb(receiver().worker(), &recv_buf[0],
-                                   size, ucp_dt_make_contig(1), message,
-                                   !err_handling ? rtag_complete_cb :
-                                   rtag_complete_err_handling_cb);
+        ucp_request_param_t recv_param = {};
+        recv_param.op_attr_mask        = UCP_OP_ATTR_FIELD_CALLBACK;
+        recv_param.cb.recv             = reinterpret_cast
+                                         <ucp_tag_recv_nbx_callback_t>(
```

Comment:
why need reinterpret_cast? because of user_data?

Reply:
yes